### PR TITLE
Migrate to Swift 3.0 on Xcode 8 GM

### PR DESCRIPTION
**Pull request by giginet**
I migrated code to Swift 3 on Xcode 8 GM 💪 
It seems to work well and all tests are passed on Xcode/SPM.
I PRed to master, however you should create `swift-3` branch and change the base to it.

This implementation is inspired by #69.
When #69  is merged, then this should be closed. 

Please check it 🙇

### CODE REVIEW

File: .swift-version [hunk 1]
```diff
@@ -1 +1 @@
-DEVELOPMENT-SNAPSHOT-2016-03-24-a
+3.0-GM-CANDIDATE
```
Comment (giginet) [inline .swift-version]: Use 3.0-GM-CANDIDATE to build by SPM
all of dependencies(Quick/Nimble/Result) uses this snapshot.

File: .travis.yml [hunk 2]
```diff
@@ -11,18 +11,12 @@ matrix:
     - script: set -o pipefail && xcodebuild -workspace "$TRAVIS_XCODE_WORKSPACE" -scheme "$TRAVIS_XCODE_SCHEME" test | xcpretty
       xcode_scheme: Commandant
       os: osx
-      osx_image: xcode7.3
```
Comment (giginet) [inline .travis.yml]: Drop Xcode 7.3 support 👋

File: Package.swift [hunk 3]
```diff
@@ -3,6 +3,6 @@ import PackageDescription
 let package = Package(
   name: "Commandant",
   dependencies: [
-    .Package(url: "https://github.com/antitypical/Result.git", "2.0.0")
+    .Package(url: "https://github.com/antitypical/Result.git", "3.0.0-alpha.4")
```
Comment (giginet) [inline Package.swift]: Unfortunately, SPM could not fetch from specific branch.
So it's using different Result version between Xcode and SPM.

Comment (mdiep) [inline Package.swift]: I just released Result 3.0.0 today. 😄

Comment (giginet) [inline Package.swift]: This problem will solves if Result will be released 3.0.0.

Comment (mdiep) [inline Package.swift]: https://github.com/antitypical/Result/releases/tag/3.0.0

Comment (giginet) [inline Package.swift]: @mdiep Sounds great!  I'll try it

File: .swift-version (hunk 1 above)
Comment (NachoSoto) [inline .swift-version]: Nice

File: Commandant.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -408,8 +425,26 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = D00CCDFC1A20719500109F8C /* Release.xcconfig */;
 			buildSettings = {
+				CLANG_WARN_BOOL_CONVERSION = YES;
```
Comment (NachoSoto) [inline Commandant.xcodeproj/project.pbxproj]: These should already by in `xcconfigs`: can you remove them?

File: Commandant.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -446,6 +482,7 @@
 				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/../Frameworks @loader_path/Frameworks";
 				PRODUCT_BUNDLE_IDENTIFIER = "org.carthage.$(PRODUCT_NAME:rfc1034identifier)";
 				PRODUCT_NAME = "$(TARGET_NAME)";
+				SWIFT_VERSION = 3.0;
```
Comment (NachoSoto) [inline Commandant.xcodeproj/project.pbxproj]: Can you make sure that this is defined at the project level, and not in the individual targets?

File: Sources/Commandant/Errors.swift [hunk 6]
```diff
@@ -10,9 +10,9 @@ import Foundation
 import Result
 
 #if swift(>=3)
-	public typealias ClientErrorType = ErrorProtocol
+	public typealias ClientErrorType = Error
 #else
-	public typealias ClientErrorType = ErrorType
+	public typealias ClientErrorType = ErrorProtocol
```
Comment (NachoSoto) [inline Sources/Commandant/Errors.swift]: I don't think `Swift 2.x` is supported after all these changes, so I'd just remove this.

File: Sources/Commandant/LinuxSupport.swift [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import Foundation
+
+// swift-corelibs-foundation is still written in Swift 2 API.
```
Comment (NachoSoto) [inline Sources/Commandant/LinuxSupport.swift]: LOL seriously.

File: Cartfile [hunk 8]
```diff
@@ -1 +1 @@
-github "antitypical/Result" ~> 2.1.3
+github "antitypical/Result" "3.0.0"
```
Comment (ikesyo) [inline Cartfile]: Can you please change this not to pin to a specific tag? `github "antitypical/Reault" ~> 3.0`

Comment (giginet) [inline Cartfile]: Sure.
And I also modified Package.swift

```swift
.Package(url: "https://github.com/antitypical/Result.git", majorVersion: 3, minor: 0)
```

File: Commandant.xcodeproj/project.pbxproj (hunk 5 above)
Comment (giginet) [inline Commandant.xcodeproj/project.pbxproj]: This property seems to be generated automatically when sources had been converted.
I couldn't find this on GUI. So I guess we can't change it.

**Comment by giginet** [comment]
Thank you. I fixed reviewed points.

File: Sources/Commandant/ArgumentType.swift [hunk 9]
```diff
@@ -12,21 +12,21 @@ public protocol ArgumentType {
...
+	public static func fromString(_ string: String) -> Int? {
 		return Int(string)
 	}
 }
 
 extension String: ArgumentType {
 	public static let name = "string"
 
-	public static func fromString(string: String) -> String? {
+	public static func fromString(_ string: String) -> String? {
```
Comment (mdiep) [inline Sources/Commandant/ArgumentType.swift]: These should be renamed `from(string: String)` to be in the Swift 3 style. (With an `@available` that points to the rename.)

File: Sources/Commandant/Command.swift [hunk 10]
```diff
@@ -90,15 +90,15 @@ public final class CommandRegistry<ClientError: ClientErrorType> {
...
+	public func register<C: CommandType>(_ command: C) where C.ClientError == ClientError, C.Options.ClientError == ClientError {
 		commandsByVerb[command.verb] = CommandWrapper(command)
 	}
 
 	/// Runs the command corresponding to the given verb, passing it the given
 	/// arguments.
 	///
 	/// Returns the results of the execution, or nil if no such command exists.
-	public func runCommand(verb: String, arguments: [String]) -> Result<(), CommandantError<ClientError>>? {
+	public func runCommand(_ verb: String, arguments: [String]) -> Result<(), CommandantError<ClientError>>? {
```
Comment (mdiep) [inline Sources/Commandant/Command.swift]: I think this should be renamed `run(command verb: String, arguments: [String])`

**Comment by giginet** [comment]
Sorry for my late reply.
I fixed reviewed points. Please check it.

File: Commandant.xcodeproj/project.pbxproj (hunk 5 above)
Comment (NachoSoto) [inline Commandant.xcodeproj/project.pbxproj]: It is in the GUI under "legacy Swift" or something like that.

File: Sources/Commandant/Errors.swift [hunk 11]
```diff
@@ -9,79 +9,75 @@
 import Foundation
 import Result
 
-#if swift(>=3)
-	public typealias ClientErrorType = ErrorProtocol
-#else
-	public typealias ClientErrorType = ErrorType
-#endif
+public typealias ClientErrorType = Error
```
Comment (NachoSoto) [inline Sources/Commandant/Errors.swift]: Let's just drop the `typealias`, it doesn't add any value anymore, especially not as `pulibc`.

File: Sources/Commandant/Option.swift [hunk 12]
```diff
@@ -109,41 +109,36 @@ extension Option: CustomStringConvertible {
...
-	precedence 150
-}
+infix operator <| : MultiplicationPrecedence
 
 /// Applies `f` to the value in the given result.
 ///
 /// In the context of command-line option parsing, this is used to chain
 /// together the parsing of multiple arguments. See OptionsType for an example.
-public func <*> <T, U, ClientError>(f: T -> U, value: Result<T, CommandantError<ClientError>>) -> Result<U, CommandantError<ClientError>> {
+public func <*> <T, U, ClientError>(f: @escaping (T) -> U, value: Result<T, CommandantError<ClientError>>) -> Result<U, CommandantError<ClientError>> {
```
Comment (NachoSoto) [inline Sources/Commandant/Option.swift]: Huh? This shouldn't be `@escaping`.

File: Commandant.xcodeproj/project.pbxproj (hunk 5 above)
Comment (giginet) [inline Commandant.xcodeproj/project.pbxproj]: OK, I found

**Comment by argon** [comment]
🙏

**Comment by ikesyo** [comment]
I'm closing this in favor of #69. Thanks for the pull request though @giginet!

**Comment by giginet** [comment]
OK, no problem.
Thank you!

**Pull request closed without merging**